src/lib/modes/mode_pad/mode_pad.h

```diff
@@ -63,16 +76,36 @@ class BOTAN_TEST_API BlockCipherModePaddingMethod {
       * virtual destructor
       */
       virtual ~BlockCipherModePaddingMethod() = default;
+
+   protected:
+      /**
+      * Applies the concrete padding to the @p last_block assuming the padding
+      * bytes should start at @p padding_start_pos within the last block.
+      *
+      * Concrete implementations of this function must ensure not to leak
+      * @p padding_start_pos via side channels.
+      */
+      virtual void apply_padding(std::span<uint8_t> last_block, size_t padding_start_pos) const {
+         BOTAN_UNUSED(last_block, padding_start_pos);
+         BOTAN_ASSERT_UNREACHABLE();
```

Comment:
I don't understand why this isn't a pure virtual

Reply:
To support `Null_Padding` the `add_padding` method (that contains the sanity checks) must be `virtual` and is overridden by `Null_Padding` as a NOOP. Defining this `apply_padding` customization point with a "useless" implementation allowed me to avoid overriding `apply_padding` in `Null_Padding`.

I agree though, that this is fairly unusual. Changed to pure virtual here and moved the `BOTAN_ASSERT_UNREACHABLE` override definition down into `Null_Padding`.